### PR TITLE
Implement keyed messages

### PR DESCRIPTION
**Pull request by alpreu**
This PR should implement keyed messaging on the PulsarTemplate (#26). 
A detailed description follows.

### CODE REVIEW

File: spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarOperations.java [hunk 1]
```diff
@@ -71,7 +71,20 @@ default MessageId send(T message, MessageRouter messageRouter) throws PulsarClie
...
+	/**
+	 * Sends a message to the specified topic in a blocking manner.
+	 * @param topic the topic to send the message to or {@code null} to send to the default topic
+	 * @param message the message to send
+	 * @param messageRouter the optional message router to use
+	 * @param typedMessageBuilderCustomizer the optional TypedMessageBuilder customizer
+	 * @return the id of the sent message
+	 * @throws PulsarClientException if an error occurs
+	 */
+	MessageId send(String topic, T message, MessageRouter messageRouter, TypedMessageBuilderCustomizer<T> typedMessageBuilderCustomizer) throws PulsarClientException;
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarOperations.java:87]: Wdyt about adding a few more combos to accommodate use of customizer w/o router (and vice-versa)?

**Current combos:**
```
- message
- message / topic
- message / router 
- message / topic / router
```
**New combos something like this:** 
``` 
- message / customizer
- message / topic / customizer
- message / customizer / router
- message / topic / customizer / router
```
The tradeoff of course is signature explosion vs. user convenience. For me, the concerns of signature explosion is lightened by the `default impls delegate to the single abstract w/ all of the params` pattern we are using.

The other tiny change I think we can make it put the customizer in front of the router in the arg list as I bet router is used less than customizer and the customizer would come directly after the payload arg (cohesive).

File: spring-pulsar/src/main/java/org/springframework/pulsar/core/TypedMessageBuilderCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.core;
+
+import java.util.function.Function;
+import org.apache.pulsar.client.api.TypedMessageBuilder;
+
+/**
+ * The function to customize a {@link TypedMessageBuilder}.
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/core/TypedMessageBuilderCustomizer.java:23]: I used an actual function in the backing issue, but that was mostly for illustration. I think we can use something similar to what Boot uses such as [here](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/DefaultKafkaProducerFactoryCustomizer.java). 

The 2 important things in there are 
* `@FunctionalInterface` marker on the interface
* Signature change slightly to `void customize(TheThingWeAreCustomizingHere<?, ?> theThing)`

**Review by onobc — CHANGES_REQUESTED** [review]
That was fast @alpreu ! Thanks, looks great. I have a few minor comments/suggestions to follow up on. Additionally,

1. Please add yourself as `@author` on all modified files. 
2. Do you mind adding some tests cases to the existing PulsarTemplateTests?

Thanks again for the contribution!

File: spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarOperations.java (hunk 1 above)
Comment (alpreu) [reply spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarOperations.java:87]: Thanks for the suggestions, and yes, having the customizer together with the payload makes sense 👍

File: spring-pulsar/src/main/java/org/springframework/pulsar/core/TypedMessageBuilderCustomizer.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.apache.pulsar.client.api.TypedMessageBuilder;
+
+/**
+ * The interface to customize a {@link TypedMessageBuilder}.
+ *
+ * @author Alexander Preuß
+ */
+@FunctionalInterface
+public interface TypedMessageBuilderCustomizer {
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/core/TypedMessageBuilderCustomizer.java:27]: Wdyt about adding the type `<T>` here and parameterize the parameter below w/ it?

**Review by onobc — CHANGES_REQUESTED** [review]
Looks great @alpreu . Thanks for the updates and adding the tests. The only comment I have is around type info to the customizer. Once we resolve that we can remove the draft status and I will get it merged in.

**Comment by alpreu** [comment]
Thank you for the feedback, I added the type to the message builder and updated the docs to reflect the changes. If it looks good to you I'll squash and rebase so you can merge :)

**Review by onobc — APPROVED** [review]

**Comment by onobc** [comment]
Looks good @alpreu . I went ahead and squash/rebase/merge. Thanks again for contribution!

**Pull request merged**